sklearn/preprocessing/_target_encoder.py

```diff
@@ -321,26 +360,88 @@ def _fit_encodings_all(self, X, y):
             dtype=np.int64,
             count=len(self.categories_),
         )
-        if self.smooth == "auto":
-            y_variance = np.var(y)
-            self.encodings_ = _fit_encoding_fast_auto_smooth(
-                X_ordinal, y, n_categories, self.target_mean_, y_variance
+        if self.target_type_ == "multiclass":
+            encodings = self._learn_multiclass_encodings(
+                X_ordinal,
+                y,
+                n_categories,
             )
+            self.encodings_ = encodings
         else:
-            self.encodings_ = _fit_encoding_fast(
-                X_ordinal, y, n_categories, self.smooth, self.target_mean_
+            self._learn_encodings(
+                X_ordinal,
+                y,
+                n_categories,
+                self.target_mean_,
+                save_encodings=True,
             )
 
         return X_ordinal, X_known_mask, y, n_categories
 
+    def _learn_encodings(self, X_ordinal, y, n_categories, target_mean, save_encodings):
+        """Learn target encodings, optionally saving as attribute or returning."""
+        if self.smooth == "auto":
+            y_variance = np.var(y)
+            encodings = _fit_encoding_fast_auto_smooth(
+                X_ordinal,
+                y,
+                n_categories,
+                target_mean,
+                y_variance,
+            )
+        else:
+            encodings = _fit_encoding_fast(
+                X_ordinal,
+                y,
+                n_categories,
+                self.smooth,
+                target_mean,
+            )
+        if save_encodings:
+            self.encodings_ = encodings
+        else:
+            return encodings
+
+    def _learn_multiclass_encodings(self, X_ordinal, y, n_categories):
+        n_features = self.n_features_in_
+        n_classes = self._label_binarizer_.classes_.shape[0]
+        reorder_index = [
+            idx
+            for start in range(n_features)
+            for idx in range(start, (n_classes * n_features), n_features)
+        ]
```

Comment:
Since the reorder_index is being used once, it can be a generator:

```suggestion
        reorder_index = (
            idx
            for start in range(n_features)
            for idx in range(start, (n_classes * n_features), n_features)
        )
```

Can there be a comment that describes the reorder index.